Expand `following` documentation in ccr overview (#39936)
elastic/elasticsearch: docs/reference/ccr/overview.asciidoc

that {ccr} does not interfere with indexing on the leader index.

[float]

> === Initiating replication

Reviewer: I'm curious about the intended change in meaning here. Is it meant to cover both configuring and starting CCR?

Reviewer 2: I guess so. The actual `create follower` and `auto follower` actions seem more associated with initiating replication (although you can configure certain things using parameters). That is why I made that change. Thoughts?

Reviewer: I find "initiating" a bit unclear, so my vote would be simpler verbs like "configuring" or "setting up", but it's not a deal breaker.

Reviewer 2: I changed it back to `configuring`

NOTE: You must also <<ccr-requirements,configure the leader index>>.

> 

Reviewer: I think it would be helpful to point out that all of those tasks can be done via Kibana too and add a link to `{kibana-ref}/working-remote-clusters.html#managing-cross-cluster-replication[Managing {ccr}]`

Reviewer 2: I have added an additional note. Let me know if that looks good.

Reviewer: I added a new commit that mentions Kibana directly in the bullet points, since I couldn't add a suggestion on that section.

> When you initiate replication either manually or through a auto-follow pattern, the

Reviewer 3: `through a auto` -> `through an auto`

follower index is created on the local cluster. Once the follower index is created,

> the <<remote-recovery, remote recovery>> process bulk transfers all of the Lucene

Reviewer 3: `transfers` -> `copies` (and maybe drop the word `bulk`)

files from the remote cluster to the local cluster.

> When initiating following manually using the API, by default the recovery process

Reviewer: ```suggestion
By default, if you initiate following manually (by using {kib} or the create follower API), the recovery process
```

> will be asynchronous in relationship to the

Reviewer: ```suggestion
is asynchronous in relationship to the
```

> {ref}/ccr-put-follow.html[create follower request]. The request will return prior

Reviewer: ```suggestion
{ref}/ccr-put-follow.html[create follower request]. The request returns before
```

> to the <<remote-recovery, remote recovery>> process being completed. If you

Reviewer: ```suggestion
the <<remote-recovery, remote recovery>> process completes. If you
```

would like to wait on the process to complete, you can use the

> `wait_for_active_shards` parameter to do so.

Reviewer: ```suggestion
`wait_for_active_shards` parameter. For example:
```

Reviewer 2: I'm not sure `For example` makes sense here. That code is commented out. It only sets up the test for the falling behind example. Wait_for_active_shards is defined and shown on the api page.

Reviewer: You're right, nevermind!

//////////////////////////

[source,js]
--------------------------------------------------
PUT /follower_index/_ccr/follow?wait_for_active_shards=1
{
  "remote_cluster" : "remote_cluster",
  "leader_index" : "leader_index"
}
--------------------------------------------------
// CONSOLE
// TESTSETUP
// TEST[setup:remote_cluster_and_leader_index]

[source,js]
--------------------------------------------------
POST /follower_index/_ccr/pause_follow
--------------------------------------------------
// CONSOLE
// TEARDOWN

//////////////////////////

[float]

> === The mechanics of replication

Reviewer: I would suggest putting this content into a separate page (akin to the "How it works" pages in the other sections), since it's quite low level for an overview. That can be done in a separate PR if necessary.

Reviewer 2: I have added a task to the meta issue #35975

follower shard task resumes sending read requests when the write buffer no
longer exceeds its configured limits.

NOTE: The intricacies of how operations are replicated from the leader are

> governed by settings that can be configured in the

Reviewer: ```suggestion
governed by settings that you can configure when you create the follower index in {kib}
```

> {ref}/ccr-put-follow.html[create follower request].

Reviewer: ```suggestion
or by using the {ref}/ccr-put-follow.html[create follower API].
```

on the follower shard task if your initial configuration is not suitable for
your use case.

[float]

> === Leader index retaining operations for replication

Reviewer: If the  "mechanics of replication" section is turned into a separate page, I think this info about how to retain operations might be a good fit there too.

Reviewer 2: I have added a task to the meta issue #35975

generated when the operation was initially performed. Lucene segment files are
occasionally merged in order to optimize searches and save space. When these
merges occur, it is possible for operations associated with deleted or updated
documents to be pruned during the merge. When the follower requests the sequence
number for a pruned operation, the process will fail due to the operation missing
on the leader.

This scenario is not possible in an append-only workflow. As documents are never
deleted or updated, the underlying operation will not be pruned.

Elasticsearch attempts to mitigate this potential issue for update workflows using
a Lucene feature called soft deletes. When a document is updated or deleted, the
underlying operation is retained in the Lucene index for a period of time. This
period of time is governed by the `index.soft_deletes.retention_lease.period`
setting which defaults to `12 hours`.

When a follower initiates the index following, it acquires a retention lease from
the leader. This informs the leader that it should not allow a soft delete to be
pruned until either the follower indicates that it has received the operation or

> the lease expires after `12 hours`. It is valuable to have monitoring in place to

Reviewer 3: I find the formatting on "12 hours" odd. Is there a reason that you chose to format it this way, as opposed to "twelve hours" (without being formatted as code) or `12h`. Another concern I have is if we change the defaults from `12h` to some other value and we would have to maintain all the pages in the docs. Maybe the best option is to add docs for the `index.soft_deletes.retention_lease.period` setting (or link to the CCR requirements page) and let the default be only specified there?

Reviewer 2: I linked the requirements page.

replication issue prior to the lease expiring so that the problem can be remedied
before the follower falls fatally behind.

[float]

> === Remedying a follower that has fallen behind

Reviewer: I think this content could actually be helpful in the Troubleshooting section (i.e. https://www.elastic.co/guide/en/elastic-stack-overview/master/troubleshooting.html)

Reviewer 2: I have added a task to the meta issue #35975

If a follower falls sufficiently behind a leader that it can no longer replicate

> operations this can be detected using the

Reviewer: I presume this exception would be visible in Kibana too?:

```suggestion
operations this can be detected in {kib} or by using the
```

Reviewer 2: I have made this change. @cjcenizal a leader falling behind if reported as a fatal exception in the stats api. Is that shown on the Kibana ui?

Reviewer 4: I'm going to tag in @jen-huang and @sebelga to answer your question, since I'm on PTO today and focusing on stack upgrade testing for the rest of the week.

Reviewer 5: @tbrooks8 It doesn't appear we show `indicies[].fatal_exception` in the UI. I've created a bug ticket for this: https://github.com/elastic/kibana/issues/33628

{ref}/ccr-get-follow-stats.html[get follow stats API]. It will be reported as a
`indices[].fatal_exception`.

> In order to restart the follower, the following process must be paused, the follower

Reviewer: ```suggestion
In order to restart the follower, you must pause the following process, close the follower
```

> index closed, and the create follower API called again.

Reviewer: ```suggestion
index, and the create follower index again. For example:
```

["source","js"]
----------------------------------------------------------------------
POST /follower_index/_ccr/pause_follow

POST /follower_index/_close

PUT /follower_index/_ccr/follow?wait_for_active_shards=1
{
  "remote_cluster" : "remote_cluster",
  "leader_index" : "leader_index"
}
----------------------------------------------------------------------
// CONSOLE

> Calling the create follower API is a destructive action. All of the existing Lucene

Reviewer: Updating to apply to Kibana too:
```suggestion
Re-creating the follower index is a destructive action. All of the existing Lucene
```

> segment files will be deleted on the follower cluster. The

Reviewer: ```suggestion
segment files are deleted on the follower cluster. The
```

> <<remote-recovery, remote recovery>> process will be used to copy the Lucene segment

Reviewer: ```suggestion
<<remote-recovery, remote recovery>> process copies the Lucene segment
```

> files from the leader again. After the follower index has been reinitialized, the

Reviewer: ```suggestion
files from the leader again. After the follower index initializes, the
```

> following process will be started again.

Reviewer: ```suggestion
following process starts again.
```
